### ISSUE TITLE
How should notifications be handled in the sync API?

### ISSUE THREAD

**Issue by sfackler** [sfackler/rust-postgres#404]

**Comment by riendegris** [comment]
Hello,

I am trying to use this new notification API in my application. This application uses `warp::sse` to  forward notifications from Postgres to the web client. It uses `tokio_postgres` to store things in the database.

So I am trying to poll for notifications at regular intervals. If there is a notification, I will send the first one. 

But this code, although it compiles, does not run as planned... When the client requests this stream, the runtime panics:

```
thread 'tokio-runtime-worker' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thre
ad is being used to drive asynchronous tasks.
```

Here is the function that  builds a stream

```
fn get_stream() -> Result<
   impl Stream<Item = Result<impl ServerSentEvent + Send + 'static, warp::Error>> + Send + Sync + 'static,
   error::Error,
> {
      Ok(time::interval(Duration::from_secs(1))
        .then(|_| {
             let mut client = postgres::Client::connect( "postgresql://user:pwd@postgres:5432/db", postgres::NoTls,).expect("connection");
             client.execute("LISTEN documents", &[]).unwrap();
             let notif = client .notifications() .iter() .collect::<Vec<_>>() .expect("notifications");
             drop(client);
             future::ready(notif)
            })
            .filter_map(|ns| {
                if ns.len() == 0 {
                    future::ready(None)
                } else {
                    future::ready(Some(Ok(( sse::event(String::from(ns[0].channel())), sse::data(String::from(ns[0].payload())),))))
                }
            }))
    }
```

For good measure.... I tried a full asynchronous solution before that. I tried to use a code similar to the `tokio_postgres/tests/test/main.rs`, using a `client.batch_execute("LISTEN documents").await?;`

But for some reason, this statement never reach the database (Looking at the postgres logs in debug mode, I should see the database receiving this query (`2020-03-31 19:08:15.247 CEST [9992] LOG:  statement: LISTEN document` if I use a C API). But there is no such line in the logs.

Would someone kindly point me in the right direction to solve this problem please?

I know the paint is still fresh on this API.

**Comment by sfackler** [comment]
You cannot use the blocking client in a nonblocking context.

I don't know why you aren't seeing the query logged - other than that do you not see the effect you would expect?

**Comment by riendegris** [comment]
For the async version, which uses the same technique as in the tokio_postgres test.main.rs (mpsc unbounded channel), I don't see the `LISTEN` in the postgres, and I don't see the notifications received in the rx stream. If I put a `debug!` statement after (see code below), I don't see it with a `RUST_LOG=debug`. (I see, however, `DEBUG tokio_postgres::simple_query > executing statement batch: LISTEN documents`).

```
        client
            .batch_execute("LISTEN documents")
            .await
            .context(error::DBError)?;
                                                                                                                                                                                                                                                                                                                                               
        debug!("After LISTEN documents");
        drop(client);

        Ok(rx.filter_map(|m| match m {
            AsyncMessage::Notification(n) => {
                debug!("Received notification on channel: {}", n.channel());
                future::ready(Some(Ok((
                    sse::event(String::from(n.channel())),
                    sse::data(String::from(n.payload())),
                ))))
            }
            _ => {
                debug!("Received something on channel."); 
                future::ready(None)
            }
        }))
```

**Comment by sfackler** [comment]
Is the connection running?

**Comment by riendegris** [comment]
Probably.... the code before the one above looks like

```
        let (client, mut connection) =
            connect_raw("postgresql://user:pwd@postgres:5432/journal")
                .await
                .context(error::DBError)?;

        let (tx, rx) = futures::channel::mpsc::unbounded();
        let stream = stream::poll_fn(move |cx| connection.poll_message(cx)).map_err(|e| panic!(e));
        let connection = stream.forward(tx).map(|r| r.expect("stream forward"));
        tokio::spawn(connection);
        debug!("Spawned dedicated connection for postgres notifications");
```

So the connection is spawned... But I am not testing it before executing a statement with the client.

**Comment by riendegris** [comment]
If I put a `debug!("connection is {:?}", client.is_closed())` after the `client.batch_execute(...).await?;`, it never gets executed.
Is this because when polled it constantly return a Poll::Pending? Or is this because it's a future, and nobody is asking for its resolution?
I'm sorry for my lack of understanding of the background...

**Comment by sfackler** [comment]
That would mean the batch_execute isn't completing.

**Comment by riendegris** [comment]
This is probably something I am missing on my side (as I use your library in other ways and it never fails me).
I've tried to use `query` instead of `batch_execute`, but that's not improving.
Could I monitor something over the network (like wireshark) to see if something is missing?
I'm also sending you the link to the project, maybe I've missed an important assumption?
[github repo](https://github.com/riendegris/journal). This repo might be a better venue for exchanging about my issue, which does not seem related to the topic here "How should notifications be handled in the sync API?"

Anyway, thanks for your help!

**Comment by sfackler** [comment]
You're calling that function in a futures::executor::block_on, which doesn't have any hooks to deal with tokio IO: https://github.com/riendegris/journal/blob/master/src/main.rs#L136

**Comment by sfackler** [comment]
Implemented in #588
